runtime/src/registrar.rs

```diff
@@ -0,0 +1,381 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Main parachains logic. For now this is just the determination of which validators do what.
+
+use rstd::prelude::*;
+#[cfg(any(feature = "std", test))]
+use rstd::marker::PhantomData;
+use parity_codec::{Encode, Decode};
+#[cfg(any(feature = "std", test))]
+use sr_primitives::{StorageOverlay, ChildrenStorageOverlay};
+use sr_primitives::{
+	weights::{SimpleDispatchInfo, DispatchInfo}, transaction_validity::ValidTransaction,
+	traits::{Hash as HashT, StaticLookup, DispatchError, SignedExtension}
+};
+#[cfg(feature = "std")]
+use srml_support::storage::hashed::generator;
+use srml_support::{
+	decl_storage, decl_module, ensure,
+	StorageValue, StorageMap, Dispatchable, dispatch::Result,
+	traits::{Get, Currency}
+};
+use system::{self, ensure_root, ensure_signed};
+use primitives::parachain::{
+	Id as ParaId, CollatorId, Scheduling, LOWEST_USER_ID, OnSwap, Info as ParaInfo, ActiveParas
+};
+use crate::parachains;
+
+/// Parachain registration API.
+pub trait Registrar<AccountId> {
+	/// Create a new unique parachain identity for later registration.
+	fn new_id() -> ParaId;
+
+	/// Register a parachain with given `code` and `initial_head_data`. `id` must not yet be registered or it will
+	/// result in a error.
+	fn register_para(
+		id: ParaId,
+		info: ParaInfo,
+		code: Vec<u8>,
```

Comment:
A few of the functions have arguments like `(..., code: Vec<u8>, initial_head_data: Vec<u8>, ...)`.

Could we declare and then import as necessary some type aliases like:
```
pub type CodeHash = T::Hash;
pub type InitialHeadData = T::Hash;
```

And then change all those function arguments to `(..., code: CodeHash, initial_head_data: InitialHeadData, ...)`.